[TEST-FAILURE] JoinStressTest.testTCPIPJoinWithManyNodesMultipleGroups

```
java.lang.AssertionError: null
    at org.junit.Assert.fail(Assert.java:86)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertNotNull(Assert.java:621)
    at org.junit.Assert.assertNotNull(Assert.java:631)
    at com.hazelcast.cluster.JoinStressTest.testJoinWithManyNodesMultipleGroups(JoinStressTest.java:169)
    at com.hazelcast.cluster.JoinStressTest.testTCPIPJoinWithManyNodesMultipleGroups(JoinStressTest.java:120)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/406/testReport/junit/com.hazelcast.cluster/JoinStressTest/testTCPIPJoinWithManyNodesMultipleGroups/

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterServiceImpl.java
@@ -544,6 +544,11 @@ public final class ClusterServiceImpl implements ClusterService, ConnectionListe
 
         if (node.getMasterAddress() != null) {
             sendMasterAnswer(joinMessage.getAddress());
+        } else {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Received a master question from " + joinMessage.getAddress()
+                        + ", but this node is not master itself or doesn't have a master yet!");
+            }
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
@@ -156,6 +156,10 @@ public class TcpIpJoiner extends AbstractJoiner {
                         claimingMaster = false;
                         return;
                     }
+                } else {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest("Cannot claim myself as master! Will try to connect a possible master...");
+                    }
                 }
 
                 claimingMaster = false;
@@ -167,7 +171,11 @@ public class TcpIpJoiner extends AbstractJoiner {
     }
 
     private boolean claimMastership(Collection<Address> possibleAddresses) {
-        logger.finest("Claiming myself as master node!");
+        if (logger.isFinestEnabled()) {
+            Set<Address> votingEndpoints = new HashSet<Address>(possibleAddresses);
+            votingEndpoints.removeAll(blacklistedAddresses.keySet());
+            logger.finest("Claiming myself as master node! Asking to endpoints: " + votingEndpoints);
+        }
         claimingMaster = true;
         Collection<Future<Boolean>> responses = new LinkedList<Future<Boolean>>();
         for (Address address : possibleAddresses) {
@@ -211,6 +219,9 @@ public class TcpIpJoiner extends AbstractJoiner {
     private boolean isThisNodeMasterCandidate(Collection<Address> possibleAddresses) {
         int thisHashCode = node.getThisAddress().hashCode();
         for (Address address : possibleAddresses) {
+            if (isBlacklisted(address)) {
+                continue;
+            }
             if (node.connectionManager.getConnection(address) != null) {
                 if (thisHashCode > address.hashCode()) {
                     return false;

